### ISSUE TITLE
Jedi struggles with pybind11 modules

### ISSUE THREAD

**Issue by vinbradev** [davidhalter/jedi#1388]
We want to offer a completion for our pybind11 modules using Jedi 0.15.1. I don't understand why it works sometime and sometimes not. Let's start with a simple example:

```cpp
PYBIND11_MODULE( completiontest, m )
{
  m.def( "myFunc", []( int intVal, std::string strVal ) {
    std::string intAsStr  = std::to_string( intVal );
    std::string resultStr = "intVal: " + intAsStr + ", strVal: " + strVal;
    return resultStr;
  } );   
}
```

In Python:
```python
import completiontest # module is available and loaded successfully
completiontest.myF # function name will successfully be completed
completiontest.myFunc( # we get no completion for the parameters of the function
completiontest.myFunc( 1, "TEST"). # we get no completions for the return type of the function
```

I don't know what the problem is, but I've tried to help Jedi understand the function by giving a docstring. By default, pybind11 automatically generates and prepends a signature to the docstring of a function, see: [Generating documentation using Sphinx](https://pybind11.readthedocs.io/en/stable/advanced/misc.html#generating-documentation-using-sphinx)
For our function this is the generated `__doc__:  myFunc(arg0: int, arg1: str) -> str`
Now Jedi is able to return the parameters of the function. Unfortunately the generated docstring does not contain the right parameter names. So Jedi rightly returns arg0 and arg1. But we still get no completions for the return type of the function. So I continued to follow the docstring approach and defined my own docstring (the pybind11 docstring generation must be deactivated):

```cpp
m.def( "myFunc",
         []( int intVal, std::string strVal ) {
           std::string intAsStr  = std::to_string( intVal );
           std::string resultStr = "intVal: " + intAsStr + ", strVal: " + strVal;
           return resultStr;
         },
         R"-(
             .. py:function:: send_message(sender, recipient, message_body, [priority=1])
               :param str sender: The person sending the message
               :param str recipient: The recipient of the message
               :param str message_body: The body of the message
               :param priority: The priority of the message, can be a number 1-5
               :type priority: integer or None
               :return: the message id
               :rtype: str
               :raises ValueError: if the message_body exceeds 160 characters
               :raises TypeError: if the message_body is not a basestrinpug
    )-" );
```

As you can see the dosctring has absolutely nothing to do with the function. But Jedi is able to complete and return all the contents of the docstring. Even the return type can be completed. Now I have tried to transfer the findings to my actual problem. As soon as you use object-oriented-code the docstring approach does not work anymore. Source: [Creating bindings for a custom type](https://pybind11.readthedocs.io/en/stable/classes.html)

```cpp
struct Pet {
    Pet(const std::string &name) : name(name) { }
    void setName(const std::string &name_) { name = name_; }
    const std::string &getName() const { return name; }

    std::string name;
};

  py::class_<Pet>( m, "Pet" )
    .def( py::init<const std::string&>() )
    .def( "setName", &Pet::setName,
          R"-(
                .. py:function:: setName(name)
                   :param str name: Pets name
      )-" )
    .def( "getName", &Pet::getName,
          R"-(
             .. py:function:: getName()
               :return: the pets name
               :rtype: str
               :raises ValueError: Doesn't matter
               :raises TypeError: Doesn't matter
      )-" );
```

Jedi is able to complete the module, the object Pet and its functions. But we get no information about the parameters and we are not able to complete the return type of the function getName(). The curious thing is that the Jedi Interpreter is able to complete it. But only if a docstring is given.

So our questions are:

1. Why is Jedi not able to complete the pybind11 modules without any help of the docstrings?
2. Why sometimes Jedi is able to offer signatures and completions using the docstring and sometimes don't?
3. Why can Jedi interpret the dosctrings for objects when using the interpreter, but not when using Script.completions()?
4. What would be the right approach for completing pybind11 modules, functions and objects?

@davidhalter This is the issue you suggested on stackoverflow, but hopefully with more helpful information: [How to set __doc__ correctly for the autocompletion of C++ libraries](https://stackoverflow.com/questions/56739155/how-to-set-doc-correctly-for-the-autocompletion-of-c-libraries)

**Comment by davidhalter** [comment]
Hey @vinbradev 

Thanks for the writeup. The general idea is that using docstrings for returns/parameters is very fragile. It was a feature we used a lot before Python 3.5. Starting with Python3.5, argument clinic made it into the Python core and `inspect.signature` returned way better values.

So is it an option for you guys to get `inspect.signature(pybind11_func)` working? Because if that works, Jedi will pick it up and everything will be great. I'm not exactly sure how to do it, because I have never done it. But according to the code it looks like you have to use a `types.BuiltinFunctionType` function and define `__text_signature__` like this:

```
$ python3.7
>>> str.split.__text_signature__
'($self, /, sep=None, maxsplit=-1)'
>>> import inspect
>>> inspect.signature(str.split)
<Signature (self, /, sep=None, maxsplit=-1)>

>>> callable.__text_signature__
'($module, obj, /)'
>>> inspect.signature(callable)
<Signature (obj, /)>

>>> inspect.signature(foo)
<Signature (a, b)>
>>> foo.__signature__ = inspect.signature(callable)
>>> inspect.signature(foo)
<Signature (obj, /)>

```

It's also possible to just use `your_object.__signature__` (the last case), where you need to properly use an `inspect.Signature` object. In that case there's no need for a proper "builtin".

Does that help? I plan on removing the support for those "docstring signatures". It's just not where Python is heading and it's kind of hard to parse.

**Comment by vinbradev** [comment]
@davidhalter thank you very much for the clarification and advices. It looks like pybind11 does not support `__signature` and `__text_signature__`: [Set the __text_signature__ attribute of callables](https://github.com/pybind/pybind11/issues/945)
Maybe it will in the future.

Now I saw that you recently added stub files support in Jedi. I did a quick test and it seems to work for my pybind11 modules. But this means I have to generate stub files after every change in my module. In the best case, I can automate that. Otherwise, I fear that this will be forgotten during development.

After the latest findings I believe that this could be the only way to get a proper completion for pybind11 modules with Jedi. What do you think? Or is there maybe a better/more practical way to achieve this using python/CPython built-in functionality?

**Comment by davidhalter** [comment]
Stubs are definitely the way to go, if you want proper type inference. I still think working with `__signature__` should be the goal (however you achieve that), but that's kind of a different issue.

Stubs work and they are great. Feel free to use them :)

**Comment by davidhalter** [comment]
Closing, because IMO issues are solved here (regarding Jedi). I know that it's probably not easy to write extensions and set `__signature__` and docstrings properly. Unfortunately I don't really have experience and cannot help you in that field.
